### PR TITLE
Adding mixin class for ease saving, uploading, downloading (as discussed in issue #9).

### PR DESCRIPTION
**Pull request by vasudevgupta7**
This PR will add feature as discussed in issue #9.

## Example
```python
from huggingface_hub import ModelHubMixin

class MyModel(nn.Module, ModelHubMixin):
   def __init__(self, **kwargs):
      super().__init__()
      self.config = kwargs.pop("config", None)
      self.layer = ...
   def forward(self, ...):
      return ...

model = MyModel()

# saving model to local directory & pushing to hub
model.save_pretrained("mymodel", upload_to_hub=True, config={"act": "gelu"})

# initiatizing model & loading it from trained-weights
model = MyModel.from_pretrained("username/mymodel@main")
```

## Note
 - In case user want to save some specific layers as per his needs, he can simple overwrite `._save_pretrained()` method.
 - Supporting all kind of `PyTorch` model.
 - `tf2` not supported currently

### CODE REVIEW

**Comment by vasudevgupta7** [comment]
Shall I integrate model uploading directly in `.save_pretrained(path, model_id="username/mymodel", upload_to_hub=True)` method instead of making separate method `.upload_to_hub(model_id="username/mymodel")`? I feel this looks better.

Please put your views on this.

**Comment by julien-c** [comment]
Hi @vasudevgupta7 thanks for the PR! 

I have no more bandwidth this week but plan on actively looking at this early next week!

File: src/huggingface_hub/hub_mixin.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            subprocess.run(["git", "checkout", branch], stdout=subprocess.PIPE)
+
+        with open(".git/config", "r") as f:
+            git_config = f.read().split()
+        if (PREFIX+model_id) not in git_config:
+            subprocess.run(["git", "remote", "add", "origin", PREFIX+model_id], stdout=subprocess.PIPE)
+            subprocess.run(["git", "fetch", "origin"], stdout=subprocess.PIPE)
+            if branch in os.listdir(".git/refs/remotes/origin"):
+                subprocess.run(["git", "merge", f"origin/{branch}"], stdout=subprocess.PIPE)
+        os.chdir("../")
```
Comment (julien-c) [inline src/huggingface_hub/hub_mixin.py:221]: I think the git part itself will be in a `Repository` class (see #7) that exposes a number of helpers to do git-related operations from Python

So this mixin will be mostly glue-code gluing together other classes.

How does this sound?

Comment (vasudevgupta7) [reply src/huggingface_hub/hub_mixin.py:221]: Yeah, that will make code more cleaner. I will update this, after #7 will be merged to the repo.

File: src/huggingface_hub/hub_mixin.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+    def _push_git(self, save_directory, track_extensions, commit_message, branch):
+
+        os.chdir(save_directory)
+        subprocess.run(["git-lfs", "track"]+track_extensions, stdout=subprocess.PIPE)
```
[the remark this answers is not in the document]
Comment (vasudevgupta7) [reply src/huggingface_hub/hub_mixin.py:226]: If user wants to add file corresponding to some other extension lets say `.pt`, then it will allow them to do that. Currently `.gitattributes` is not having this extension which is very common.  Also, if user wants to commit to some branch `ver1` (different from `main`) without relating commit histories of 2 branches, then its is possible only if he add `.gitattributes` manually by this command.

The reason, I am fetching and merging (instead of cloning repo) is that, it will give freedom to name local directory whatever they want (will help if they doing epoch pushing during training) and still keeping the hub repo name to `model_id`.

File: src/huggingface_hub/hub_mixin.py (hunk 1 above)
Comment (vasudevgupta7) [reply src/huggingface_hub/hub_mixin.py:221]: @julien-c Looks like author of PR #7 is slightly busy. Since this PR is dependent on that, I would like to take over his PR. Is it possible?

**Comment by julien-c** [comment]
Ooops @vasudevgupta7 I wanted to reach out to let you know that we had merged the Repository PR. But I guess you saw that :)

Will take a look at your updated PR with @philschmid and others, soon. Thanks!

File: src/huggingface_hub/hub_mixin.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            model = cls(**kwargs)
+
+        state_dict = torch.load(model_file, map_location=map_location)
+        model.load_state_dict(state_dict, strict=strict)
+        model.eval()
+
+        return model
+
+    @staticmethod
+    def upload_to_hub(weights_directory: str, model_id: str, **kwargs):
```
Comment (julien-c) [inline src/huggingface_hub/hub_mixin.py:168]:
```suggestion
    def push_to_hub(weights_directory: str, model_id: str, **kwargs):
```

to be consistent with `Repository`'s method?

File: src/huggingface_hub/hub_mixin.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        """
+
+        commit_message = kwargs.pop("commit_message", "add model")
+        organization = kwargs.pop("organization", None)
+        private = kwargs.pop("private", None)
+
+        token = HfFolder.get_token()
+        repo_url = HfApi().create_repo(token, model_id, organization=organization, private=private, repo_type=None, exist_ok=True)
+        repo = Repository(weights_directory, clone_from=repo_url, use_auth_token=token)
+        repo.push_to_hub(commit_message=commit_message)
```
Comment (julien-c) [inline src/huggingface_hub/hub_mixin.py:182]: maybe return this, to return the url to the commit?
```suggestion
        return repo.push_to_hub(commit_message=commit_message)
```

**Comment by julien-c** [comment]
> Shall I integrate model uploading directly in .save_pretrained(path, model_id="username/mymodel", upload_to_hub=True) method instead of making separate method .upload_to_hub(model_id="username/mymodel")? I feel this looks better.

Replying late to this but yeah I feel like it'd be nice to expose both.

File: src/huggingface_hub/hub_mixin.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                Directory having model-weights & config.
+        """
+
+        commit_message = kwargs.pop("commit_message", "add model")
+        organization = kwargs.pop("organization", None)
+        private = kwargs.pop("private", None)
+
+        token = HfFolder.get_token()
+        repo_url = HfApi().create_repo(token, model_id, organization=organization, private=private, repo_type=None, exist_ok=True)
+        repo = Repository(weights_directory, clone_from=repo_url, use_auth_token=token)
```
Comment (julien-c) [inline src/huggingface_hub/hub_mixin.py:181]: does this work if the local directory is already a git clone of the model in question, i.e. does it work to call this multiple times (say, after each epoch of a training)?

Ultimately we'll need some tests too I guess

Comment (vasudevgupta7) [reply src/huggingface_hub/hub_mixin.py:181]: No, it won't work then. Shall I get that working?

Comment (julien-c) [reply src/huggingface_hub/hub_mixin.py:181]: yes that would be awesome. It shouldn't require too many changes actually (maybe a tiny change to `Repository` to handle the case where we instantiate with both a local_dir that is a git repo, AND a `clone_from` that's the same remote)

Let me know if you need help!

**Review by sgugger — COMMENTED** [review]
Thanks for the PR! I'm not entirely sure adding a complex logic in the default `save_pretrained` and `from_pretrained` is desirable as users that need such a logic will probable need to overload those methods anyway (for instance transformers will definitely need to, or the API assumes the model is initialized with a config when it exists but some library might use a class method instead). So I would limit those two methods to the base save of the model weights and document they can be overloaded to add more logic.

The push_to_hub on the other hand is absolutely awesome to have!

File: src/huggingface_hub/hub_mixin.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                config_url = hf_hub_url(model_id, filename=CONFIG_NAME, revision=revision)
+                config_file = cached_download(config_url, cache_dir=cache_dir, force_download=force_download, proxies=proxies, resume_download=resume_download, local_files_only=local_files_only, use_auth_token=use_auth_token)
+            except:
+                config_file = None
+
+        if name in os.listdir():
+            print("LOADING weights from local directory")
+            model_file = os.path.join(name, PYTORCH_WEIGHTS_NAME)
+        else:
+            model_url = hf_hub_url(model_id, filename=PYTORCH_WEIGHTS_NAME, revision=revision)
```
Comment (LysandreJik) [inline src/huggingface_hub/hub_mixin.py:151]: In other implementations, the revision can also be obtained from a split of the model_id on `@`. I think it would make sense to also support that here, so that one may do:
```py
model = MyModel.from_pretrained("username/mymodel@branch")
```
As it is already the case in several libraries.

See [asteroid](https://github.com/asteroid-team/asteroid/blob/3c6abcfc0f5dafb8cf446488f340dedd4b46124d/asteroid/utils/hub_utils.py#L59-L61) or [pyannote](https://github.com/julien-c/pyannote-audio/blob/eb4a021db13c63d8d7f1230ec1c4462db780e480/pyannote/audio/core/model.py#L684-L686)

**Review by LysandreJik — COMMENTED** [review]
This looks like a great addition!

**Comment by julien-c** [comment]
@vasudevgupta7 let us know when you feel like this is ready to squash/merge and I'll make a final review. Thanks again

File: src/huggingface_hub/hub_mixin.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+        model_id = pretrained_model_name_or_path
+        strict = kwargs.pop("strict", True)
+        map_location = kwargs.pop("map_location", torch.device("cpu"))
+        force_download = kwargs.pop("force_download", False)
+        resume_download = kwargs.pop("resume_download", False)
+        proxies = kwargs.pop("proxies", None)
+        use_auth_token = kwargs.pop("use_auth_token", None)
+        cache_dir = kwargs.pop("cache_dir", None)
+        local_files_only = kwargs.pop("local_files_only", False)
```
Comment (theo-m) [inline src/huggingface_hub/hub_mixin.py:135]: these should be explicit in the function signature, they will be separated from the model kwargs for free, e.g.
```python
    @classmethod
    def from_pretrained(
        cls, 
        pretrained_model_name_or_path: str, 
        strict:bool=true, 
        map_location:str="cpu", 
        <other args>, 
        **kwargs
    ):

```

File: src/huggingface_hub/hub_mixin.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+                config_file = cached_download(
+                    config_url,
+                    cache_dir=cache_dir,
+                    force_download=force_download,
+                    proxies=proxies,
+                    resume_download=resume_download,
+                    local_files_only=local_files_only,
+                    use_auth_token=use_auth_token,
+                )
+            except:
```
Comment (theo-m) [inline src/huggingface_hub/hub_mixin.py:157]: there is no CI yet, so it wasn't caught, but we shouldn't have bare excepts, you should try to find out what errors you can expect from the try block

File: src/huggingface_hub/hub_mixin.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            repo_url (:obj:`str`, `optional`):
+                Specify this in case you want to push to existing repo in hub.
+            organization (:obj:`str`, `optional`):
+                Organization in which you want to push your model.
+            private (:obj:`bool`, `optional`):
+                private: Whether the model repo should be private (requires a paid huggingface.co account)
+            commit_message (:obj:`str`, `optional`, defaults to :obj:`add model`):
+                Message to commit while pushing
+        """
+        repo_url = kwargs.pop("repo_url", None)
```
Comment (theo-m) [inline src/huggingface_hub/hub_mixin.py:207]: same observation re: kwargs

File: src/huggingface_hub/repository.py [hunk 10]
```diff
@@ -159,14 +159,19 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
...
+                try:
+                    subprocess.run(
+                        ["git", "remote", "add", "origin", repo_url],
+                        stderr=subprocess.PIPE,
+                        stdout=subprocess.PIPE,
+                        check=True,
+                        encoding="utf-8",
+                        cwd=self.local_dir,
+                    )
+                except:
```
Comment (theo-m) [inline src/huggingface_hub/repository.py:172]: bare except

File: src/huggingface_hub/repository.py [hunk 11]
```diff
@@ -184,14 +189,18 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
...
+                try:
+                    subprocess.run(
+                        "git checkout origin/main -ft".split(),
+                        stderr=subprocess.PIPE,
+                        stdout=subprocess.PIPE,
+                        encoding="utf-8",
+                        check=True,
+                        cwd=self.local_dir,
+                    )
+                except:
```
Comment (theo-m) [inline src/huggingface_hub/repository.py:201]: bare except

File: src/huggingface_hub/repository.py [hunk 12]
```diff
@@ -318,7 +327,7 @@ def git_pull(self, rebase: Optional[bool] = False):
         """
         git pull
         """
-        args = "git pull".split()
+        args = "git pull origin main".split()
```
Comment (theo-m) [inline src/huggingface_hub/repository.py:330]: are these changes strictly necessary?

**Review by theo-m — COMMENTED** [review]
shallow review, leaving actual logic to main reviewers

File: src/huggingface_hub/repository.py (hunk 12 above)
Comment (vasudevgupta7) [reply src/huggingface_hub/repository.py:330]: Nope, I will remove it.

File: src/huggingface_hub/hub_mixin.py (hunk 8 above)
Comment (vasudevgupta7) [reply src/huggingface_hub/hub_mixin.py:157]: Actually these `try` will fail when `push_to_hub` is called over `git` initialized repo. I will figure out some other way to do things and remove `except` statements.

File: src/huggingface_hub/hub_mixin.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+            >>> model.push_to_hub("mymodel", "model-1") # Pushing model-weights to hf-hub
+
+            >>> # Downloading weights from hf-hub & model will be initialized from those weights
+            >>> model = MyModel.from_pretrained("username/mymodel@main")
+        """
+
+    def save_pretrained(
+        self,
+        save_directory: str,
+        config: dict = None,
```
Comment (patrickvonplaten) [inline src/huggingface_hub/hub_mixin.py:40]:
```suggestion
        config: Optional[dict] = None,
```

File: src/huggingface_hub/hub_mixin.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        push_to_hub: bool = False,
+        model_id: str = None,
```
Comment (patrickvonplaten) [inline src/huggingface_hub/hub_mixin.py:42]:
```suggestion
        model_id: Optional[str] = None,
```

**Review by patrickvonplaten — APPROVED** [review]
Great PR!

**Comment by vasudevgupta7** [comment]
Hoping everything works now. Sorry for the delay.

**Comment by julien-c** [comment]
Looks good! 

Does anyone tagged here have any last comments? @LysandreJik and @philschmid maybe?

Otherwise I'll merge and push a release soon.

**Review by LysandreJik — APPROVED** [review]
LGTM!

**Review by philschmid — APPROVED** [review]
LGTM too! Thank you for creating a test. Looking forward to using this!🚀

**Comment by julien-c** [comment]
I've removed torch as a hard dependency, and please note that I don't think the tests can actually pass given that they are pushing to your userspace @vasudevgupta7 (not in a staging env) and the CI env of course doesn't have your git credentials.

I'll merge like this but would be nice to set up CI for this at some point 

@LysandreJik By the way do you know how to activate GitHub Actions for pull requests from forks?

**Comment by julien-c** [comment]
Thanks for the contribution @vasudevgupta7 ❤️

**Comment by LysandreJik** [comment]
Yes, setting it up can be done like it was done here: https://github.com/huggingface/transformers/pull/10383/files. I'm opening a PR with that proposal now!

**Comment by vasudevgupta7** [comment]
Glad it's merged finally 🤩. Thank you @julien-c, @philschmid, @LysandreJik, @patrickvonplaten, @theo-m, @sgugger for reviewing and approving.

**Comment by vasudevgupta7** [comment]
@julien-c there is small error in the example you shared in release. It should be `push_to_hub` instead of `upload_to_hub`. Sorry, I forgot to update PR description. I changed it now.

**Comment by julien-c** [comment]
Oops, no big deal, but I will update the release notes: https://github.com/huggingface/huggingface_hub/releases/tag/v0.0.7

**Comment by julien-c** [comment]
[link to 🐦 ](https://twitter.com/julien_c/status/1372613568244371461)
